lightning-invoice/src/utils.rs

```diff
@@ -255,4 +365,135 @@ mod test {
+				let payment_preimage = nodes[1].node.get_payment_preimage(payment_hash, payment_secret).unwrap();
+				(payment_preimage, payment_hash, payment_secret)
+			}
+		};
+		let channels_1 = nodes[1].node.list_channels();
+		let channels_2 = nodes[2].node.list_channels();
+		let mut route_hints = channels_1.iter().map(|e| (nodes[1].node.get_phantom_scid(), nodes[1].node.get_our_node_id(), e)).collect::<Vec<(u64, PublicKey, &ChannelDetails)>>();
+		for channel in channels_2.iter() {
+			route_hints.push((nodes[2].node.get_phantom_scid(), nodes[2].node.get_our_node_id(), &channel));
+		}
+		let invoice = ::utils::create_phantom_invoice::<EnforcingSigner, &test_utils::TestKeysInterface>(Some(payment_amt), "test".to_string(), payment_hash, payment_secret, &route_hints, &nodes[1].keys_manager, Currency::BitcoinTestnet).unwrap();
+
+		assert_eq!(invoice.min_final_cltv_expiry(), MIN_FINAL_CLTV_EXPIRY as u64);
+		assert_eq!(invoice.description(), InvoiceDescription::Direct(&Description("test".to_string())));
+		assert_eq!(invoice.route_hints().len(), 2);
+		assert!(!invoice.features().unwrap().supports_basic_mpp());
+
+		let payee = Payee::from_node_id(invoice.recover_payee_pub_key())
+			.with_features(invoice.features().unwrap().clone())
+			.with_route_hints(invoice.route_hints());
+		let params = RouteParameters {
+			payee,
+			final_value_msat: invoice.amount_milli_satoshis().unwrap(),
+			final_cltv_expiry_delta: invoice.min_final_cltv_expiry() as u32,
+		};
+		let first_hops = nodes[0].node.list_usable_channels();
+		let network_graph = node_cfgs[0].network_graph;
+		let logger = test_utils::TestLogger::new();
+		let scorer = test_utils::TestScorer::with_fixed_penalty(0);
+		let route = find_route(
+			&nodes[0].node.get_our_node_id(), &params, network_graph,
+			Some(&first_hops.iter().collect::<Vec<_>>()), &logger, &scorer,
+		).unwrap();
+		let (payment_event, fwd_idx) = {
+			let mut payment_hash = PaymentHash([0; 32]);
+			payment_hash.0.copy_from_slice(&invoice.payment_hash().as_ref()[0..32]);
+			nodes[0].node.send_payment(&route, payment_hash, &Some(invoice.payment_secret().clone())).unwrap();
+			let mut added_monitors = nodes[0].chain_monitor.added_monitors.lock().unwrap();
+			assert_eq!(added_monitors.len(), 1);
+			added_monitors.clear();
+
+			let mut events = nodes[0].node.get_and_clear_pending_msg_events();
+			assert_eq!(events.len(), 1);
+			let fwd_idx = match events[0] {
+				MessageSendEvent::UpdateHTLCs { node_id, .. } => {
+					if node_id == nodes[1].node.get_our_node_id() {
+						1
+					} else { 2 }
+				},
+				_ => panic!("Unexpected event")
+			};
+			(SendEvent::from_event(events.remove(0)), fwd_idx)
+		};
+		nodes[fwd_idx].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &payment_event.msgs[0]);
+		commitment_signed_dance!(nodes[fwd_idx], nodes[0], &payment_event.commitment_msg, false, true);
+		expect_pending_htlcs_forwardable!(nodes[fwd_idx]);
```

Comment:
This is a bit confusing because `expect_pending_htlcs_forwardable` actually calls `process_pending_htlc_forwards` twice to ensure we don't hit some edge-cases. Given that we want to explicitly test for the behavior that we have to call it twice here, maybe something like this is a bit clearer. Note that afterwards I believe you can replace the PaymentReceived match with an `expect_payment_received`.

```
diff --git a/lightning-invoice/src/utils.rs b/lightning-invoice/src/utils.rs
index 850d20d23..54aa91287 100644
--- a/lightning-invoice/src/utils.rs
+++ b/lightning-invoice/src/utils.rs
@@ -456,15 +456,19 @@ mod test {
                };
                nodes[fwd_idx].node.handle_update_add_htlc(&nodes[0].node.get_our_node_id(), &payment_event.msgs[0]);
                commitment_signed_dance!(nodes[fwd_idx], nodes[0], &payment_event.commitment_msg, false, true);
-               expect_pending_htlcs_forwardable!(nodes[fwd_idx]);
+
+               // Note that we have to "forward pending HTLCs" twice before we see the PaymentReceived as
+               // this "emulates" the payment taking two hops, providing some privacy to make phantom node
+               // payments "look real" by taking more time.
+               expect_pending_htlcs_forwardable_ignore!(nodes[fwd_idx]);
+               nodes[fwd_idx].node.process_pending_htlc_forwards();
+               expect_pending_htlcs_forwardable_ignore!(nodes[fwd_idx]);
+               nodes[fwd_idx].node.process_pending_htlc_forwards();
+
                let payment_preimage_opt = if user_generated_pmt_hash { None } else { Some(payment_preimage) };
                let events = nodes[fwd_idx].node.get_and_clear_pending_events();
-               assert_eq!(events.len(), 2);
+               assert_eq!(events.len(), 1);
                match events[0] {
-                       Event::PendingHTLCsForwardable { .. } => { },
-                       _ => panic!("Unexpected event"),
-               }
-               match events[1] {
                        Event::PaymentReceived { payment_hash: ref hash, ref purpose, amt } => {
                                assert_eq!(*hash, payment_hash);
                                assert_eq!(amt, payment_amt);
```